document/lib/Pdf/Document/FontDict.hs

```diff
@@ -222,3 +231,85 @@ loadUnicodeCMap pdf fontDict =
+      fontStretch <- optional "FontStretch" nameValue' fd
+      fontWeight <- optional "FontWeight" intValue fd
+      flags <- required "Flags" int64Value fd
+      fontBBox <- optional "FontBBox"
+        (join . fmap (either (const Nothing) Just . rectangleFromArray) . arrayValue) fd
+      italicAngle <- required "ItalicAngle" realValue fd
+      ascent <- optional "Ascent" realValue fd
+      descent <- optional "Descent" realValue fd
+      leading <- optional "Leading" realValue fd
+      capHeight <- optional "CapHeight" realValue fd
+      xHeight <- optional "XHeight" realValue fd
+      stemV <- optional "StemV" realValue fd
+      stemH <- optional "StemH" realValue fd
+      avgWidth <- optional "AvgWidth" realValue fd
+      maxWidth <- optional "MaxWidth" realValue fd
+      missingWidth <- optional "MissingWidth" realValue fd
+      charSet <- optional "CharSet" (fmap decodeUtf8 . stringValue) fd
+
+      return $ Just $ FontDescriptor
+        { fdFontName = fontName
+        , fdFontFamily = fontFamily
+        , fdFontStretch = fontStretch
+        , fdFontWeight = fontWeight
+        , fdFlags = flags
+        , fdFontBBox = fontBBox
+        , fdItalicAngle = italicAngle
+        , fdDescent = descent
+        , fdAscent = ascent
+        , fdLeading = leading
+        , fdCapHeight = capHeight
+        , fdXHeight = xHeight
+        , fdStemV = stemV
+        , fdStemH = stemH
+        , fdAvgWidth = avgWidth
+        , fdMaxWidth = maxWidth
+        , fdMissingWidth = missingWidth
+        , fdCharSet = charSet
+        }
+  where
+    required = requiredInDict "FontDescriptor"
+    optional = optionalInDict "FontDescriptor"
+    nameValue' = fmap (decodeUtf8 . Name.toByteString) . nameValue
+
+-- | Parse a value from a required field of a dictionary. This will
+-- raise an exception if a) the field is not present or b) the field
+-- value has a false type.
+requiredInDict :: String -- ^ a context for a failure notice
+               -> Name   -- ^ name of dictionary field
+               -> (Object -> Maybe a) -- ^ function for type-casting the object
+               -> Dict                -- ^ the dictionary
+               -> IO a
+requiredInDict context key typeFun =
+  join .
+  sure . (`notice` (context ++ ": " ++ msg ++ " should exist")) .
+  fmap (sure . (`notice` (context ++ ": " ++ msg ++ " type failure")) . typeFun) .
+  HashMap.lookup key
+  where
+    msg = Text.unpack $ decodeUtf8 $ Name.toByteString key
+
+-- | Parse a value from an optional field of a dictionary. This will
+-- raise an exception if the field value has a false type.
+optionalInDict :: String -> Name -> (Object -> Maybe a) -> Dict -> IO (Maybe a)
+optionalInDict context key typeFun =
+  maybe (return Nothing) (liftM Just) .
+  fmap (sure . (`notice` (context ++ ": " ++ msg ++ " type failure")) . typeFun) .
+  HashMap.lookup key
+  where
+    msg = Text.unpack $ decodeUtf8 $ Name.toByteString key
```

Comment:
`key` is a `Name`, and it may not be UTF-8 encoded. And `decodeUtf8` may throw an exception in such case.
It's OK to try to interpret it as UTF-8 encoded string, but I'd use `decodeUtf8With ignore` to make sure we never throw an exception here. The same for the `requiredInDict`.

Reply:
Very good point! I've replaced `decodeUtf8` with `decodeUtf8With ignore` as you suggest.